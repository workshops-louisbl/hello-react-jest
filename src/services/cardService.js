export function getCards() {
  return fetch('/card').then(res => res.json())
}

export function destroyCard(id) {
  return fetch(`/card/${id}`, { method: 'DELETE' })
}

export function saveCard(card) {
  return card.id ? updateCard(card) : createCard(card)
}

export function createCard(card) {
  return fetch('/card', {
    method: 'POST',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify(card)
  }).then(res => res.json())
}

function updateCard(card) {
  return fetch(`/card/${card.id}`, {
    method: 'PUT',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify(card)
  }).then(res => res.json())
}
