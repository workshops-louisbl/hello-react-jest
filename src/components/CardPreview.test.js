import React from 'react';
import { CardPreview } from "./CardPreview";
import { render, fireEvent, act } from "@testing-library/react";

import { destroyCard } from "../services/cardService";
jest.mock("../services/cardService.js");

function setup() {
  const term = "This is a test term";
  const definition = "This is the definition of a test term";
  const utils = render(<CardPreview term={term} definition={definition} />);

  return {
    term,
    definition,
    utils
  }
}
it("renders correctly", () => {
  const { utils } = setup();

  const fragment = utils.asFragment();

  expect(fragment).toMatchSnapshot();
});

it("shows term correctly", () => {
  const {term, utils} = setup();

  // assert
  expect(utils.getByText(term)).toBeInTheDocument();
})

it("has a button to show definition", () => {
  const {utils} = setup();

  expect(utils.getByText("show back")).toBeVisible()
})

it("shows definition when it flips the card", () => {
  const {term, definition, utils} = setup();

  const showButton = utils.getByRole("button", {name: "show back"});
  fireEvent.click(showButton);

  expect(utils.getByText(definition)).toBeInTheDocument();
  expect(utils.queryByText(term)).not.toBeInTheDocument();
})

it("flips the card back to the term", () => {
  const {term, definition, utils} = setup();

  const showBackButton = utils.getByRole("button", {name: "show back"});
  fireEvent.click(showBackButton);

  const showFrontButton = utils.getByRole("button", {name: "show front"});
  fireEvent.click(showFrontButton);

  expect(utils.getByText(term)).toBeInTheDocument();
  expect(utils.queryByText(definition)).not.toBeInTheDocument();
})

it("can delete the card", async () => {
  window.confirm = jest.fn(() => {
    return true;
  });
  destroyCard.mockResolvedValue();

  const idProps = "0";
  const onRemoveProps = jest.fn();

  const utils = render(<CardPreview id={idProps} onRemove={onRemoveProps} />);

  const deleteButton = utils.getByRole("button", {name: "delete"});

  await act(async () => {
    fireEvent.click(deleteButton);
  })

  expect(window.confirm).toBeCalled();
  expect(destroyCard).toBeCalled();
  expect(onRemoveProps).toBeCalledWith(idProps);
})

it("can edit the card", () => {
  const {term, definition, utils} = setup();

  const editButton = utils.getByRole("button", {name: "edit"});

  act(() => {
    fireEvent.click(editButton);
  })

  expect(editButton).not.toBeInTheDocument();
})
