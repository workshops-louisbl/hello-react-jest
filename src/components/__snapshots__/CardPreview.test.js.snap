// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="tile "
  >
    <h4
      class="cardTerm"
    >
      This is a test term
    </h4>
    <div
      class="cardButtons"
    >
      <button
        class="tertiary"
        type="button"
      >
        show back
      </button>
      <div>
        <button
          class="secondary"
          type="button"
        >
          edit
        </button>
        <button
          class="secondary danger"
          type="button"
        >
          delete
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
